In [1]:
import dataPreparationUtils as dpu
from scipy.stats import gamma, lognorm, pareto
from rpy2.robjects.packages import importr
import rpy2.robjects as ro
import pandas as pd
import numpy as np
import os
from rpy2.robjects import numpy2ri

In [2]:
import numpy as np
import pandas as pd
import itertools
import os
import readDataV2
import latexCodeGenV2

In [3]:
FILE_DIR = "small"
nExperimentReptition = 10
randomSeed = 20220222
trueValue = 0.005
dataDistributions = ['gamma', 'lognorm', 'pareto']
dataSizes = [500]

In [3]:
np.linspace(0, 0.9, 10)

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

## 1. Read data

- Tail probability estimation with single threshold
- Tail probability estimation with multiple thresholds
- Quantile estimation with single threshold
- Quantile estimation with multiple thresholds

In [3]:
# tb1 = readDataV2.tableOne().groupby(by=['dataDistribution','dataSize','percentageLHS','thresholdPercentage'])
# tb2 = readDataV2.tableTwo().groupby(by=['dataDistribution','dataSize','quantitleValue','thresholdPercentage'])
# tb31 = readDataV2.tableThreeOne().groupby(by=['dataDistribution','dataSize','percentageLHS','thresholdPercentage'])
# tb32 = readDataV2.tableThreeTwo().groupby(by=['dataDistribution','dataSize','percentageLHS','thresholdPercentage'])
# tb33 = readDataV2.tableThreeThree().groupby(by=['dataDistribution','dataSize','quantitleValue','thresholdPercentage'])
# tb34 = readDataV2.tableThreeFour().groupby(by=['dataDistribution','dataSize','quantitleValue','thresholdPercentage'])
# tb4 = readDataV2.tableFour().groupby(by=['dataDistribution','dataSize','percentageLHS','thresholdPercentage'])
tb5 = pd.read_csv(os.path.join(FILE_DIR, 'tableFive_bayesian.csv')).groupby(by=['Data Source', 'nData', 'percentageLHS'])


## Latex Code Generation

## Table 1

In [5]:
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHS = 0.99
thresholdPercentage = 0.70
targetColumns = ['(0,CHI2)','(1,CHI2)','(2,CHI2)','(0,KS)','(1,KS)','(2,KS)']
text_file = open("./robustTailTable/table1.tex", "w")
text_file.write(latexCodeGenV2.getTableOne(tb1, dataSources, nDatas, percentageLHS, thresholdPercentage, targetColumns, 
                   "Tail probability estimation under different constraint settings. ",
                   "tb1_tpe", 0.8).strip())
text_file.close()

## Table 2

In [4]:
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHS = 0.99
thresholdPercentage = 0.70
targetColumns = ['(0,CHI2)','(1,CHI2)','(2,CHI2)']
text_file = open("./robustTailTable/table2.tex", "w")
text_file.write(latexCodeGenV2.getTableTwo(tb2, dataSources, nDatas, percentageLHS, thresholdPercentage, targetColumns, 
                   "Quantile estimation under different constraint settings. ",
                   "tb2_qe", 0.75).strip())
text_file.close()

## Table 3

In [8]:
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHS = 0.99
thresholdPercentages = 0.6
targetColumns = ['(2,CHI2)']
text_file = open("./robustTailTable/table3_tpe.tex", "w")
text_file.write(latexCodeGenV2.getTableThree(tb31, tb32, dataSources, nDatas, percentageLHS, thresholdPercentages, targetColumns, 
                   "Tail probablity estimation under different cutoff threshold(s).",
                   "tb3_tpe", 0.75).strip())
text_file = open("./robustTailTable/table3_qe.tex", "w")
text_file.write(latexCodeGenV2.getTableThree(tb33, tb34, dataSources, nDatas, percentageLHS, thresholdPercentages, targetColumns, 
                   "Quantile estimation under different cutoff threshold(s). ",
                   "tb3_qe", 0.75, True).strip())
text_file.close()

## Table 4

In [9]:
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHSs = np.linspace(0.9, 0.99, 10).tolist()
thresholdPercentage = 0.7
targetColumns = ['(2,CHI2)','(2,KS)']
text_file = open("./robustTailTable/table4.tex", "w")
text_file.write(latexCodeGenV2.getTableFour(tb4, dataSources, nDatas, percentageLHSs, thresholdPercentage, targetColumns, 
                   "Tail probability estimation under different objective functions. ",
                   "tb4_tpe_{:}".format(thresholdPercentage),0.6).strip())
text_file.close()

## Table 5:

## Peak over threshold

In [26]:
file_dir = 'large_table5_pot'
df = pd.DataFrame()
for percentageLHS in ["0.9", "0.95", "0.99"]:
    for dataSource in ["gamma","lognorm", "pareto"]:
        df = pd.concat([df, pd.read_csv(f'{file_dir}/table5_{percentageLHS}_{dataSource}.csv')], axis=0)
tb5 =  df.groupby(by=['Data Source', 'nData', 'percentageLHS'])

In [27]:
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHSs = [0.9, 0.95, 0.99]
targetColumns = ['Upper Bound']
text_file = open("./robustTailTable/table5_pot.tex", "w")
text_file.write(latexCodeGenV2.getTableFive(tb5, dataSources, nDatas, percentageLHSs, targetColumns,
                   "Tail probability estimation using Peak-Over-Threshold.",
                   "tb5_tpe_pot",0.8).strip())
text_file.close()

## Peak over threshold with gof threshold

In [29]:
file_dir = 'large_table5_pot_bt'
df = pd.DataFrame()
for percentageLHS in ["0.9", "0.95", "0.99"]:
    for dataSource in ["gamma","lognorm", "pareto"]:
        df = pd.concat([df, pd.read_csv(f'{file_dir}/table5_{percentageLHS}_{dataSource}.csv')], axis=0)
tb5 =  df.groupby(by=['Data Source', 'nData', 'percentageLHS'])

In [30]:
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHSs = [0.9, 0.95, 0.99]
targetColumns = ['Upper Bound']
text_file = open("./robustTailTable/table5_pot_bt.tex", "w")
text_file.write(latexCodeGenV2.getTableFive(tb5, dataSources, nDatas, percentageLHSs, targetColumns,
                   "Tail probability estimation using Peak-Over-Threshold.",
                   "tb5_tpe_pot_bt",0.8).strip())
text_file.close()

## Profile likelihood method

In [14]:
file_dir = 'large_table5_pl'
df = pd.DataFrame()
for percentageLHS in ["0.9", "0.95", "0.99"]:
    for dataSource in ["gamma","lognorm", "pareto"]:
        df = pd.concat([df, pd.read_csv(f'{file_dir}/table5_{percentageLHS}_{dataSource}.csv')], axis=0)
tb5 =  df.groupby(by=['Data Source', 'nData', 'percentageLHS'])

In [16]:
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHSs = [0.9, 0.95, 0.99]
targetColumns = ['Upper Bound']
text_file = open("./robustTailTable/table5_w_pl.tex", "w")
text_file.write(latexCodeGenV2.getTableFive(tb5, dataSources, nDatas, percentageLHSs, targetColumns,
                   "Tail probability estimation using profile likelihood.",
                   "tb5_tpe_profile_likelihood",0.8).strip())
text_file.close()


## Bayesian method

In [11]:
file_dir = 'large_table5_bayesian'
df = pd.DataFrame()
for percentageLHS in ["0.9", "0.95", "0.99"]:
    for dataSource in ["gamma","lognorm", "pareto"]:
        df = pd.concat([df, pd.read_csv(f'{file_dir}/table5_{percentageLHS}_{dataSource}.csv')], axis=0)
tb5 =  df.groupby(by=['Data Source', 'nData', 'percentageLHS'])

In [12]:
import numpy as np
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHSs = [0.9, 0.95, 0.99]
targetColumns = ['Upper Bound']
text_file = open("./robustTailTable/table5_w_bayesian.tex", "w")
text_file.write(latexCodeGenV2.getTableFive(tb5, dataSources, nDatas, percentageLHSs, targetColumns,
                   "Tail probability estimation using Bayesian.",
                   "tb5_tpe_bayesian",0.8).strip())
text_file.close()

## Probability-weighted moment methods

In [7]:
import pandas as pd

In [8]:
file_dir = 'large_table5_pwm'
df = pd.DataFrame()
for percentageLHS in ["0.9", "0.95", "0.99"]:
    for dataSource in ["gamma","lognorm", "pareto"]:
        df = pd.concat([df, pd.read_csv(f'{file_dir}/table5_{percentageLHS}_{dataSource}.csv')], axis=0)
tb5 =  df.groupby(by=['Data Source', 'nData', 'percentageLHS'])

In [10]:
import numpy as np
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [500]
percentageLHSs = [0.9, 0.95, 0.99]
targetColumns = ['Upper Bound']
text_file = open("./robustTailTable/table5_w_pwm.tex", "w")
text_file.write(latexCodeGenV2.getTableFive(tb5, dataSources, nDatas, percentageLHSs, targetColumns,
                   "Tail probability estimation using probability-weighted moment.",
                   "tb5_tpe_pwm",0.8).strip())
text_file.close()

## Scarce Data

In [66]:
import pandas as pd
import numpy as np


In [71]:
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [30]
percentageLHSs = [0.9, 0.95, 0.99]
targetColumns = ['Upper Bound']
df_print = []
for data_src in dataSources:
    for plhs in percentageLHSs:
        df_each = tb5.get_group((data_src, 30, plhs))
        re = df_each['Upper Bound'].div(df_each['True Value'])
        re_mean, re_std = re.mean(), re.std()
        cp = df_each['Upper Bound'] > df_each['True Value']
        ub_mean, ub_std = df_each['Upper Bound'].mean(), df_each['Upper Bound'].std()
        cp_mean, cp_std = cp.mean(), cp.std()
        df_print.append([data_src, plhs, f"{re_mean:.3f}({1.96*re_std/np.sqrt(30):.3f})", f"{ub_mean:.2e}({1.96*ub_std/np.sqrt(30):.2e})", f"{cp_mean:.3f}({1.96*cp_std/np.sqrt(30):.3f})"])
df_print = pd.DataFrame(df_print, columns=["src", "lhs", "re", "ub", "cp"])
print("optimization-based method")
display(df_print)

pl


src   lhs             re                  ub            cp
0    gamma  0.90   2.678(0.187)  1.34e-02(9.34e-04)  1.000(0.000)
1    gamma  0.95   3.335(0.208)  1.67e-02(1.04e-03)  1.000(0.000)
2    gamma  0.99   9.403(0.498)  4.70e-02(2.49e-03)  1.000(0.000)
3  lognorm  0.90   2.959(0.199)  1.48e-02(9.96e-04)  0.995(0.025)
4  lognorm  0.95   3.938(0.269)  1.97e-02(1.35e-03)  1.000(0.000)
5  lognorm  0.99  12.453(0.635)  6.23e-02(3.18e-03)  1.000(0.000)
6   pareto  0.90   3.601(0.236)  1.80e-02(1.18e-03)  1.000(0.000)
7   pareto  0.95   5.050(0.273)  2.53e-02(1.37e-03)  1.000(0.000)
8   pareto  0.99  18.877(0.770)  9.44e-02(3.85e-03)  1.000(0.000)

In [60]:
tb5 = output.groupby(by=['Data Source', 'nData', 'percentageLHS'])

In [74]:
import pandas as pd
import latexCodeGenV2

for method in ['pot_bt', 'pl', 'bayesian', 'pwm']:
    file_dir = f'results/table5_abstract/n30_{method}'
    df = pd.DataFrame()
    for dataSource in ["gamma","lognorm", "pareto"]:
        df = pd.concat([df, pd.read_csv(f'{file_dir}/table5_{0.99}_{dataSource}.csv')], axis=0)
    tb5 =  df.groupby(by=['Data Source', 'nData', 'percentageLHS'])
    
    dataSources = ["gamma","lognorm", "pareto"]
    nDatas = [30]
    percentageLHSs = [0.9, 0.95, 0.99]
    targetColumns = ['Upper Bound']
    df_print = []
    for data_src in dataSources:
        for plhs in percentageLHSs:
            df_each = tb5.get_group((data_src, 30, plhs))
            re = df_each['Upper Bound'].div(df_each['True Value'])
            re_mean, re_std = re.mean(), re.std()
            cp = df_each['Upper Bound'] > df_each['True Value']
            ub_mean, ub_std = df_each['Upper Bound'].mean(), df_each['Upper Bound'].std()
            cp_mean, cp_std = cp.mean(), cp.std()
            df_print.append([data_src, plhs, f"{re_mean:.3f}({1.96*re_std/np.sqrt(30):.3f})", f"{ub_mean:.2e}({1.96*ub_std/np.sqrt(30):.2e})", f"{cp_mean:.3f}({1.96*cp_std/np.sqrt(30):.3f})"])
    df_print = pd.DataFrame(df_print, columns=["src", "lhs", "re", "ub", "cp"])
    print(method)
    display(df_print)
    # print(
    # latexCodeGenV2.getTableFive(tb5, dataSources, nDatas, percentageLHSs, targetColumns,
    #                             "Tail probability estimation using probability-weighted moment.",
    #                             f"tb5_tpe_{method}_scarce",0.8))
    

pot_bt


src   lhs            re                  ub            cp
0    gamma  0.90  1.199(0.278)  6.00e-03(1.39e-03)  0.670(0.169)
1    gamma  0.95  1.563(0.533)  7.82e-03(2.66e-03)  0.680(0.167)
2    gamma  0.99  3.097(1.342)  1.55e-02(6.71e-03)  0.580(0.177)
3  lognorm  0.90  1.328(0.266)  6.64e-03(1.33e-03)  0.765(0.152)
4  lognorm  0.95  1.939(0.467)  9.69e-03(2.33e-03)  0.800(0.143)
5  lognorm  0.99  3.239(1.381)  1.62e-02(6.91e-03)  0.615(0.175)
6   pareto  0.90  1.612(0.242)  8.06e-03(1.21e-03)  0.880(0.117)
7   pareto  0.95  2.305(0.735)  1.15e-02(3.67e-03)  0.880(0.117)
8   pareto  0.99  3.239(1.005)  1.62e-02(5.02e-03)  0.690(0.166)

pl


src   lhs            re                  ub            cp
0    gamma  0.90  1.411(0.343)  7.06e-03(1.71e-03)  0.805(0.142)
1    gamma  0.95  2.027(0.760)  1.01e-02(3.80e-03)  0.725(0.160)
2    gamma  0.99  2.644(1.268)  1.32e-02(6.34e-03)  0.600(0.176)
3  lognorm  0.90  1.504(0.263)  7.52e-03(1.31e-03)  0.855(0.126)
4  lognorm  0.95  2.283(0.615)  1.14e-02(3.08e-03)  0.795(0.145)
5  lognorm  0.99  2.717(1.234)  1.36e-02(6.17e-03)  0.625(0.174)
6   pareto  0.90  1.874(0.555)  9.37e-03(2.77e-03)  0.890(0.112)
7   pareto  0.95  2.371(0.708)  1.19e-02(3.54e-03)  0.860(0.124)
8   pareto  0.99  3.041(0.835)  1.52e-02(4.17e-03)  0.765(0.152)

bayesian


src   lhs            re                  ub            cp
0    gamma  0.90  1.390(0.283)  6.95e-03(1.42e-03)  0.805(0.142)
1    gamma  0.95  1.807(0.348)  9.04e-03(1.74e-03)  0.865(0.123)
2    gamma  0.99  3.137(0.579)  1.57e-02(2.89e-03)  0.910(0.103)
3  lognorm  0.90  1.319(0.264)  6.60e-03(1.32e-03)  0.795(0.145)
4  lognorm  0.95  1.602(0.339)  8.01e-03(1.70e-03)  0.840(0.132)
5  lognorm  0.99  2.698(0.493)  1.35e-02(2.47e-03)  0.900(0.108)
6   pareto  0.90  1.418(0.166)  7.09e-03(8.28e-04)  0.820(0.138)
7   pareto  0.95  1.677(0.267)  8.39e-03(1.34e-03)  0.925(0.094)
8   pareto  0.99  2.694(0.368)  1.35e-02(1.84e-03)  0.930(0.092)

pwm


src   lhs            re                  ub            cp
0    gamma  0.90  1.297(0.244)  6.48e-03(1.22e-03)  0.735(0.158)
1    gamma  0.95  1.543(0.572)  7.71e-03(2.86e-03)  0.665(0.169)
2    gamma  0.99  2.101(0.673)  1.05e-02(3.36e-03)  0.660(0.170)
3  lognorm  0.90  1.420(0.259)  7.10e-03(1.30e-03)  0.760(0.153)
4  lognorm  0.95  1.921(1.373)  9.61e-03(6.87e-03)  0.730(0.159)
5  lognorm  0.99  2.151(0.689)  1.08e-02(3.45e-03)  0.655(0.171)
6   pareto  0.90  1.421(0.243)  7.11e-03(1.21e-03)  0.655(0.171)
7   pareto  0.95  1.571(0.357)  7.85e-03(1.79e-03)  0.670(0.169)
8   pareto  0.99  1.601(0.562)  8.00e-03(2.81e-03)  0.555(0.178)

In [ ]:
def tableFive_w_std(groupby_object1:,
                    dataSources: list, nDatas: list, percentageLHSs: list,
                    targetColumns: list):
    trueValues = []
    contents = []
    for dataSource in dataSources:
        RelativeRatios = [[ 0 for _ in range(len(nDatas))] for _ in range(len(percentageLHSs))]
        CoverageProbabilities = [[ 0 for _ in range(len(nDatas))] for _ in range(len(percentageLHSs))]
        EstimatedUpperBounds = [[ 0 for _ in range(len(nDatas))] for _ in range(len(percentageLHSs))]
        
        for i, percentageLHS in enumerate(percentageLHSs):
            for j, nData in enumerate(nDatas):
                currKeyChoice = (dataSource, nData, round(percentageLHS, 2))
                trueValues.append(groupby_object1.get_group(currKeyChoice)['True Value'].unique()[0])
                EstimatedUpperBounds[i][j]= groupby_object1.get_group(currKeyChoice)[targetColumns].mean().tolist()
                RelativeRatios[i][j] = (groupby_object1.get_group(currKeyChoice)[targetColumns].mean().values/trueValues[-1]).tolist()
                CoverageProbabilities[i][j]=(groupby_object1.get_group(currKeyChoice)[targetColumns].values>trueValues[-1]).mean(axis=0).tolist()
        contents.append(tableFiveUnit(percentageLHSs, dataSource, EstimatedUpperBounds, RelativeRatios, CoverageProbabilities))
    ## trueValue should be same among different percentageLHS.         
    assert np.unique(trueValues).size == 1         
    return ("\\hline \n".join(contents), trueValues[0])

def getTableFive(go1:pd.core.groupby.generic.DataFrameGroupBy, 
                 dataSources: list, nDatas: list, percentageLHSs: list, targetColumns: list,
                 title:str, label: str, scalebox: float):
    content, trueValue = tableFive_w_std(go1, dataSources, nDatas, percentageLHSs, targetColumns)
    latexTable= r'''
\begin{table}[ht]
    \centering'''+\
    r'''\scalebox{{{:}}}{{'''.format(scalebox)+\
    r'''\begin{tabular}{cc|ccc}
    \toprule
    \hline''' +r'''
    Data Source & LHS Quantitle & Relative Error & Upper Bound & Coverage Probability \\\hline'''+"\n"+\
    content+\
    r'''
    \hline
    \bottomrule
    \end{tabular}}'''+\
    '''\caption{{{:}}}
    \label{{{:}}}
\end{{table}}
    '''.format(title+"The true value is {:}.".format(trueValue), label)
    return latexTable

In [122]:
output = {}
for method in ["pot_bt"]:#, "pl", "bayesian", "pwm"]:
    for dataSource in ["gamma", "lognorm", "pareto"]:
        curr_data  = []
        for rank in range(5):
            group = (dataSource, rank)
            tb5_go = tb5.get_group(group)
            ub = tb5_go['Upper Bound']
            ub = [ub.mean(), 1.96*ub.std()/np.sqrt(200)]
            rr = tb5_go['Upper Bound'].div(tb5_go['True Value'])
            rr = [rr.mean(), 1.96*rr.std()/np.sqrt(200)]
            cp = (tb5_go['Upper Bound'] >= tb5_go['True Value'])
            cp = [cp.mean(), 1.96*cp.std()/np.sqrt(200)]
            curr_data.append([ub[0], ub[1], rr[0], rr[1], cp[0], cp[1]])
        output[dataSource] = pd.DataFrame(curr_data, index=range(5), columns=["ub", "ub_std", "rr", "rr_std", "cp", "cp_std"])
        output[dataSource].index.name = 'rank'
        #     line.append(f"${rr[0]:.3f}(\pm {rr[1]:.3f})$")
        #     line.append(f"${sci_to_latex(ub[0])}(\pm {sci_to_latex(ub[1])})$")
        #     if cp[0] < 0.95:
        #         line.append(f"$\\bm{{{cp[0]:.3f}(\pm {cp[1]:.3f})}}$")
        #     else:
        #         line.append(f"${cp[0]:.3f}(\pm {cp[1]:.3f})$")
        # line = " & ".join(line) + r"\\"
        # print(line)
df = pd.concat(output,axis=0)
df.index.name = 'dataSource'
df

ub    ub_std        rr    rr_std     cp    cp_std
        rank                                                         
gamma   0     0.016007  0.001801  3.201440  0.360218  0.750  0.060028
        1     0.015939  0.002162  3.187855  0.432492  0.690  0.064116
        2     0.013969  0.002209  2.793750  0.441897  0.610  0.067620
        3     0.013867  0.002705  2.773338  0.541052  0.545  0.069040
        4     0.011026  0.002383  2.205215  0.476591  0.480  0.069270
lognorm 0     0.019584  0.002921  3.916730  0.584174  0.595  0.068051
        1     0.017900  0.002654  3.580099  0.530755  0.610  0.067617
        2     0.017189  0.002794  3.437896  0.558864  0.580  0.068425
        3     0.016605  0.002719  3.320941  0.543708  0.600  0.067921
        4     0.014523  0.003146  2.904562  0.629158  0.530  0.069200
pareto  0     0.023274  0.002554  4.654723  0.510770  0.745  0.060422
        1     0.023190  0.002588  4.637991  0.517512  0.735  0.061183
        2     0.021729  0.002481  4.345771  0.496207  0.720  0.062248
        3     0.020119  0.002380  4.023868  0.476001  0.695  0.063832
        4     0.017721  0.002369  3.544154  0.473763  0.610  0.067627

In [ ]:
# n = 50
## quantile -> seq(60%, 90%, 5%)
## other conventional 
#### OPT: 
## grid search and 70%

# real data
### -> grid search and 70%
### -> compare with conventional method. 

In [139]:
pd.DataFrame(opt_output, columns=['src', '-', "ub", "ub_std", "rr", "rr_std", "cp", "cp_std"])

src  -        ub    ub_std         rr    rr_std     cp    cp_std
0    gamma  -  0.065420  0.003763  13.084028  0.752551  0.985  0.016889
1  lognorm  -  0.090406  0.002593  18.081223  0.518521  0.995  0.009800
2   pareto  -  0.143459  0.004310  28.691734  0.862050  0.980  0.019452

In [72]:
output = []

for data_src in ['gamma', 'lognorm', 'pareto']:
    for percentage_lhs, percentage_rhs in zip(['0.9', "0.95", "0.99"],
                                              ['0.905', "0.955", "0.995"]):
        df_each = pd.read_csv(f"/home/chronotope/ws/robusttail/testResultScarce/tailProbabilityEstimation_dataDistribution={data_src}_dataSize=30_percentageLHS={percentage_lhs}_percentageRHS={percentage_rhs}_thresholdPercentage=0.7_alpha=0.05_gEllipsoidalDimension=3_randomSeed=20220222_nExperimentReptition=200.csv").set_index('Experiment Repetition Index')
        df_each['Data Source'] = data_src
        df_each['nData'] = 30
        df_each['percentageLHS'] = float(percentage_lhs)
        df_each['True Value'] = 0.005
        df_each.index.name = 'Repetition Index'
        df_each = df_each.reset_index()
        df_each = df_each.loc[:, ['Data Source',  'nData', 'percentageLHS', '(2,KS)', 'True Value', 'Repetition Index']]
        df_each = df_each.rename(columns={'(2,KS)': 'Upper Bound'})
        output.append(df_each)
output = pd.concat(output, axis=0)

In [190]:
go = output.groupby(by=['Data Source', 'rank'])

In [191]:
opt_output = []
for dataSource in ["gamma", "lognorm", "pareto"]:
    for percentage_lhs, percentage_rhs in zip(['0.9', "0.95", "0.99"],
                                              ['0.905', "0.955", "0.995"]):
        srs_ub = go.get_group((dataSource, rank))['Upper Bound']
        ubs = [srs_ub.mean(), 1.96*srs_ub.std()/np.sqrt(200)]
        rrs = srs_ub/0.005
        rrs = [rrs.mean(), 1.96*rrs.std()/np.sqrt(200)]
        cps = srs_ub>0.005
        cps = [cps.mean(), 1.96*cps.std()/np.sqrt(200)]
        output = ubs + rrs + cps
        opt_output.append([dataSource, '-'] + output)

In [192]:
pd.DataFrame(opt_output, columns=['src', '-', "ub", "ub_std", "rr", "rr_std", "cp", "cp_std"])

src  -        ub    ub_std         rr    rr_std     cp    cp_std
0    gamma  -  0.065420  0.003763  13.084028  0.752551  0.985  0.016889
1  lognorm  -  0.090406  0.002593  18.081223  0.518521  0.995  0.009800
2   pareto  -  0.143459  0.004310  28.691734  0.862050  0.980  0.019452

## Table 5. run experiments with different fixed thresholds

#### conventional method

In [1]:
import pandas as pd

In [61]:
import pandas as pd
import numpy as np
# import latexCodeGenV2

for method in ['pot', 'pl', 'bayesian', 'pwm']:
    file_dir = f'results/table5_abstract_thresholds/n30_{method}'
    df = pd.DataFrame()
    for thres_pct in [0.85]:
        for dataSource in ["gamma","lognorm", "pareto"]:
            df = pd.concat([df, pd.read_csv(f'{file_dir}/table5_{thres_pct}_{dataSource}.csv')], axis=0)
        
    tb5 =  df.groupby(by=['Data Source', 'nData', 'thres_pct'])
    dataSources = ["gamma","lognorm", "pareto"]
    nDatas = [30]
    thres_pcts = [0.6, 0.65, 0.7, 0.75, 0.8, 0.85]
    targetColumns = ['Upper Bound']
    df_print = []
    for data_src in dataSources:
        for thres_pct in thres_pcts:
            df_each = tb5.get_group((data_src, 30, thres_pct))
            re = df_each['Upper Bound'].div(df_each['True Value'])
            re_mean, re_std = re.mean(), re.std()
            cp = df_each['Upper Bound'] > df_each['True Value']
            ub_mean, ub_std = df_each['Upper Bound'].mean(), df_each['Upper Bound'].std()
            cp_mean, cp_std = cp.mean(), cp.std()
            df_print.append([data_src, thres_pct, f"{re_mean:.3f}(\pm {1.96*re_std/np.sqrt(30):.3f})", f"{ub_mean:.2e}", f"{1.96*ub_std/np.sqrt(30):.2e}", f"{cp_mean:.3f}(\pm {1.96*cp_std/np.sqrt(30):.3f})"])
            if method == 'pl' and data_src == 'gamma' and thres_pct == 0.85:
                assert False
    df_print = pd.DataFrame(df_print, columns=["src", "thres_pct", "re", "ub", "ub_std", "cp"])
    print(method)
    display(df_print)

    for thres_pct in [0.6, 0.65, 0.7, 0.75, 0.8, 0.85]:
        latex_print = [f"& ${thres_pct:.3f}$"]
        for src in ['gamma', 'lognorm', 'pareto']:
            latex_print.append("&")
            mask = (df_print['src'] == src) & (df_print['thres_pct'] == thres_pct)
            re = df_print.loc[mask, 're'].values[0]
            latex_print.append(f"${re}$")
            latex_print.append("&")
            ub = scientific_to_latex(df_print.loc[mask, 'ub'].values[0])
            ub_std = scientific_to_latex(df_print.loc[mask, 'ub_std'].values[0])
            latex_print.append(f"${ub}(\pm {ub_std})$")
            latex_print.append("&")        
            cp = df_print.loc[mask, 'cp'].values[0]
            latex_print.append(f"$\\bm{{{cp}}}$")
            
        latex_print.append("\\\\")     
        # print(" ".join(latex_print))
            
        # print(
        # latexCodeGenV2.getTableFive(tb5, dataSources, nDatas, percentageLHSs, targetColumns,
        #                             "Tail probability estimation using probability-weighted moment.",
        #                             f"tb5_tpe_{method}_scarce",0.8))
        

pot


src  thres_pct                re        ub    ub_std                cp
0     gamma       0.60  1.205(\pm 0.337)  6.02e-03  1.68e-03  0.670(\pm 0.169)
1     gamma       0.65  1.117(\pm 0.264)  5.58e-03  1.32e-03  0.620(\pm 0.174)
2     gamma       0.70  0.922(\pm 0.279)  4.61e-03  1.40e-03  0.525(\pm 0.179)
3     gamma       0.75  1.012(\pm 0.329)  5.06e-03  1.65e-03  0.515(\pm 0.179)
4     gamma       0.80  0.771(\pm 0.546)  3.85e-03  2.73e-03  0.295(\pm 0.164)
5     gamma       0.85  0.780(\pm 0.499)  3.90e-03  2.50e-03  0.265(\pm 0.158)
6   lognorm       0.60  1.382(\pm 0.253)  6.91e-03  1.26e-03  0.810(\pm 0.141)
7   lognorm       0.65  1.347(\pm 0.284)  6.74e-03  1.42e-03  0.750(\pm 0.155)
8   lognorm       0.70  1.155(\pm 0.304)  5.78e-03  1.52e-03  0.655(\pm 0.171)
9   lognorm       0.75  1.078(\pm 0.321)  5.39e-03  1.61e-03  0.540(\pm 0.179)
10  lognorm       0.80  0.780(\pm 0.357)  3.90e-03  1.79e-03  0.385(\pm 0.175)
11  lognorm       0.85  0.907(\pm 0.590)  4.54e-03  2.95e-03  0.315(\pm 0.167)
12   pareto       0.60  1.612(\pm 0.229)  8.06e-03  1.14e-03  0.890(\pm 0.112)
13   pareto       0.65  1.462(\pm 0.233)  7.31e-03  1.17e-03  0.790(\pm 0.146)
14   pareto       0.70  1.367(\pm 0.269)  6.84e-03  1.35e-03  0.715(\pm 0.162)
15   pareto       0.75  1.370(\pm 0.314)  6.85e-03  1.57e-03  0.685(\pm 0.167)
16   pareto       0.80  1.092(\pm 0.344)  5.46e-03  1.72e-03  0.550(\pm 0.178)
17   pareto       0.85  1.626(\pm 1.892)  8.13e-03  9.46e-03  0.480(\pm 0.179)

AssertionError: 

In [71]:
(df_each['Upper Bound']/0.005).std()/np.sqrt(30)*1.9

457.5500923211176

#### optimization method

In [54]:
output = []
for data_src in ['gamma', 'lognorm', 'pareto']:
    for thres_pct in [0.6, 0.65, 0.7, 0.75, 0.8, 0.85]:
        df_each = pd.read_csv(f"./testResultScarce/tailProbabilityEstimation_dataDistribution={data_src}_dataSize=30_percentageLHS=0.9_percentageRHS=0.905_thresholdPercentage={thres_pct}_alpha=0.05_gEllipsoidalDimension=3_randomSeed=20220222_nExperimentReptition=200.csv").set_index('Experiment Repetition Index')
        df_each['Data Source'] = data_src
        df_each['nData'] = 30
        df_each['thres_pct'] = float(thres_pct)
        df_each['True Value'] = 0.005
        df_each.index.name = 'Repetition Index'
        df_each = df_each.reset_index()
        df_each = df_each.loc[:, ['Data Source',  'nData', 'thres_pct', '(2,KS)', 'True Value', 'Repetition Index']]
        df_each = df_each.rename(columns={'(2,KS)': 'Upper Bound'})
        output.append(df_each)
output = pd.concat(output, axis=0)
tb5 =  output.groupby(by=['Data Source', 'nData', 'thres_pct'])
dataSources = ["gamma","lognorm", "pareto"]
nDatas = [30]
thres_pcts = [0.6, 0.65, 0.7, 0.75, 0.8, 0.85]
targetColumns = ['Upper Bound']
df_print = []
for data_src in dataSources:
    for thres_pct in thres_pcts:
        df_each = tb5.get_group((data_src, 30, thres_pct))
        re = df_each['Upper Bound'].div(df_each['True Value'])
        re_mean, re_std = re.mean(), re.std()
        cp = df_each['Upper Bound'] > df_each['True Value']
        ub_mean, ub_std = df_each['Upper Bound'].mean(), df_each['Upper Bound'].std()
        cp_mean, cp_std = cp.mean(), cp.std()
        df_print.append([data_src, thres_pct, f"{re_mean:.3f}(\pm {1.96*re_std/np.sqrt(30):.3f})", f"{ub_mean:.2e}", f"{1.96*ub_std/np.sqrt(30):.2e}", f"{cp_mean:.3f}(\pm {1.96*cp_std/np.sqrt(30):.3f})"])
df_print = pd.DataFrame(df_print, columns=["src", "thres_pct", "re", "ub", "ub_std", "cp"])
print("optimization")

display(df_print)

optimization


src  thres_pct                re        ub    ub_std                cp
0     gamma       0.60  2.558(\pm 0.149)  1.28e-02  7.46e-04  1.000(\pm 0.000)
1     gamma       0.65  2.595(\pm 0.165)  1.30e-02  8.26e-04  1.000(\pm 0.000)
2     gamma       0.70  2.678(\pm 0.187)  1.34e-02  9.34e-04  1.000(\pm 0.000)
3     gamma       0.75  2.729(\pm 0.206)  1.36e-02  1.03e-03  1.000(\pm 0.000)
4     gamma       0.80  2.806(\pm 0.388)  1.40e-02  1.94e-03  0.990(\pm 0.036)
5     gamma       0.85  2.962(\pm 0.340)  1.48e-02  1.70e-03  0.990(\pm 0.036)
6   lognorm       0.60  2.556(\pm 1.398)  1.28e-02  6.99e-03  0.995(\pm 0.025)
7   lognorm       0.65  2.858(\pm 0.172)  1.43e-02  8.62e-04  1.000(\pm 0.000)
8   lognorm       0.70  2.959(\pm 0.199)  1.48e-02  9.96e-04  0.995(\pm 0.025)
9   lognorm       0.75  2.984(\pm 0.198)  1.49e-02  9.88e-04  1.000(\pm 0.000)
10  lognorm       0.80  3.017(\pm 0.538)  1.51e-02  2.69e-03  0.990(\pm 0.036)
11  lognorm       0.85  3.123(\pm 0.294)  1.56e-02  1.47e-03  1.000(\pm 0.000)
12   pareto       0.60  3.473(\pm 0.196)  1.74e-02  9.78e-04  1.000(\pm 0.000)
13   pareto       0.65  3.504(\pm 0.207)  1.75e-02  1.03e-03  1.000(\pm 0.000)
14   pareto       0.70  3.601(\pm 0.236)  1.80e-02  1.18e-03  1.000(\pm 0.000)
15   pareto       0.75  3.646(\pm 0.241)  1.82e-02  1.21e-03  1.000(\pm 0.000)
16   pareto       0.80  3.776(\pm 0.295)  1.89e-02  1.48e-03  1.000(\pm 0.000)
17   pareto       0.85  3.741(\pm 0.419)  1.87e-02  2.10e-03  0.985(\pm 0.044)

 & $0.900$ & $1.714\ (\pm0.010)$ & $8.57\times 10^{-3}\ (\pm4.94\times 10^{-5})$ & $1.000\ (\pm0.000)$      & $1.860\ (\pm0.011)$ & $9.30\times 10^{-3}\ (\pm5.52\times 10^{-5})$ & $1.000\ (\pm0.000)$      & $2.204\ (\pm0.014)$ & $1.10\times 10^{-2}\ (\pm6.80\times 10^{-5})$ & $1.000\ (\pm0.000)$\\

In [55]:
import re as regex
def scientific_to_latex(number_str):
    match = regex.match(r"(\d*\.?\d+)[eE]([+-]?\d+)", number_str)
    base, exponent = match.groups()
    exponent = str(int(exponent))
    # Create LaTeX format string
    latex_str = f"{base} \\times 10^{{{exponent}}}"
    return latex_str

In [56]:

for thres_pct in [0.6, 0.65, 0.7, 0.75, 0.8, 0.85]:
    latex_print = [f"& ${thres_pct:.3f}$"]
    for src in ['gamma', 'lognorm', 'pareto']:
        latex_print.append("&")
        mask = (df_print['src'] == src) & (df_print['thres_pct'] == thres_pct)
        re = df_print.loc[mask, 're'].values[0]
        latex_print.append(f"${re}$")
        latex_print.append("&")
        ub = scientific_to_latex(df_print.loc[mask, 'ub'].values[0])
        ub_std = scientific_to_latex(df_print.loc[mask, 'ub_std'].values[0])
        latex_print.append(f"${ub}(\pm {ub_std})$")
        latex_print.append("&")        
        cp = df_print.loc[mask, 'cp'].values[0]
        latex_print.append(f"${cp}$")
    latex_print.append("\\\\")     
    print(" ".join(latex_print))

& $0.600$ & $2.558(\pm 0.149)$ & $1.28 \times 10^{-2}(\pm 7.46 \times 10^{-4})$ & $1.000(\pm 0.000)$ & $2.556(\pm 1.398)$ & $1.28 \times 10^{-2}(\pm 6.99 \times 10^{-3})$ & $0.995(\pm 0.025)$ & $3.473(\pm 0.196)$ & $1.74 \times 10^{-2}(\pm 9.78 \times 10^{-4})$ & $1.000(\pm 0.000)$ \\
& $0.650$ & $2.595(\pm 0.165)$ & $1.30 \times 10^{-2}(\pm 8.26 \times 10^{-4})$ & $1.000(\pm 0.000)$ & $2.858(\pm 0.172)$ & $1.43 \times 10^{-2}(\pm 8.62 \times 10^{-4})$ & $1.000(\pm 0.000)$ & $3.504(\pm 0.207)$ & $1.75 \times 10^{-2}(\pm 1.03 \times 10^{-3})$ & $1.000(\pm 0.000)$ \\
& $0.700$ & $2.678(\pm 0.187)$ & $1.34 \times 10^{-2}(\pm 9.34 \times 10^{-4})$ & $1.000(\pm 0.000)$ & $2.959(\pm 0.199)$ & $1.48 \times 10^{-2}(\pm 9.96 \times 10^{-4})$ & $0.995(\pm 0.025)$ & $3.601(\pm 0.236)$ & $1.80 \times 10^{-2}(\pm 1.18 \times 10^{-3})$ & $1.000(\pm 0.000)$ \\
& $0.750$ & $2.729(\pm 0.206)$ & $1.36 \times 10^{-2}(\pm 1.03 \times 10^{-3})$ & $1.000(\pm 0.000)$ & $2.984(\pm 0.198)$ & $1.49 \times 10^{-